### PR TITLE
[Merged by Bors] - Flatten render commands

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Speed up the render phase of rendering. Simplify the trait structure for render commands.

## Solution

 - Merge `EntityPhaseItem` into `PhaseItem` (`EntityPhaseItem::entity` -> `PhaseItem::entity`)
 - Merge `EntityRenderCommand` into `RenderCommand`.
 - Add two associated types to `RenderCommand`: `RenderCommand::ViewWorldQuery` and `RenderCommand::WorldQuery`.
 - Use the new associated types to construct two `QueryStates`s for `RenderCommandState`.
 - Hoist any `SQuery<T>` fetches in `EntityRenderCommand`s into the aformentioned two queries. Batch fetch them all at once.

## Performance
`main_opaque_pass_3d` is slightly faster on `many_foxes` (427.52us -> 401.15us)

image

The shadow pass node is also slightly faster (344.52 -> 338.24us)

image

## Future Work

 - Can we hoist the view level queries out of the core loop?

---

## Changelog
Added: `PhaseItem::entity`
Added: `RenderCommand::ViewWorldQuery` associated type.
Added: `RenderCommand::ItemorldQuery` associated type.
Added: `Draw<T>::prepare` optional trait function.
Removed: `EntityPhaseItem` trait

## Migration Guide
TODO

### CODE REVIEW

**Comment by james7132** [comment]
Saw that `QueryState::get` always calls `update_archetype` which is unnecessary when called repeatedly.

image

**Review by kurtkuehnert — COMMENTED** [review]
I do like these changes. 
I believe that everything that should be drawn can be associated with an entity. Thus, this change should not limit the user in any way. 

However, I do think that the documentation of this module is lacking. I think I am going to address this with a follow-up PR.

File: crates/bevy_render/src/render_phase/draw.rs
Comment (james7132) [reply crates/bevy_render/src/render_phase/draw.rs]: Done.

File: crates/bevy_render/src/render_phase/draw.rs
Comment (james7132) [reply crates/bevy_render/src/render_phase/draw.rs]: Done.

**Comment by james7132** [comment]
Seems like #6988 removed the `u128` equality checks and replaced them with `u32`s, which sped this up by another few percent. `many_foxes`'s main_opaque_3d stats vs the original `main` this PR was tested against.

image

**Review by kurtkuehnert — APPROVED** [review]
LGTM.

My set of follow-up documentation changes can be found in #7016.

**Review by superdump — APPROVED** [review]
Assuming everything is tested, as in all of the modified render command code paths, then LGTM.

I think @cart should take a look at this one. We originally had `PhaseItem` only, and then both agreed that `EntityPhaseItem` felt natural. I agree that it makes sense that everything that could be drawn could be associated with an `Entity` and I can't really think of a situation where that would be unreasonable. One could draw things based on a resource, but that could just as well be a single entity instead.

I definitely prefer the flatness. <3

Marking as approved...

**Comment by superdump** [comment]
Oh, it of course needs updating too.

**Comment by james7132** [comment]
Gonna hold off from resolving the merge conflicts until #7043 is merged, as it's just going to conflict again, and that PR also makes it simpler to resolve the conflicts.

**Comment by cart** [comment]
@james7132 I just `bors r+`-ed it, so it should be on main momentarily.

**Comment by james7132** [comment]
@cart should be up to date with `main` now.

**Comment by superdump** [comment]
While working on batching hacks, I noted that I would end up migrating all existing draw commands to batched draw commands. This made me think that the difference between batched and not in our implementations is simply that there is a range associated with the draw. There practically is anyway as all non-merged draw commands draw all vertices and instances `0..1`. We could flatten batched render command into render command as well and have some convenient Default impl that uses a `Range<u32>` of `0u32..1` perhaps? Or we can do it later when it's more relevant.

**Comment by inodentry** [comment]
My batching experiments involve doing away with storing `range` in phase items altogether. :)

I propose we leave anything related to batching as out of scope for this PR, and just leave the current stuff as-is.

I want to work on batching separately, and those phase item traits and struct fields are going to change anyway.

**Review by cart — APPROVED** [review]
I'm cool with this and I dig the performance improvements! It means that the phase abstraction cannot be used for non-entity driven use cases, but I'm cool with throwing that out given that we don't have any non-entity-driven use cases. And those will have different constraints anyway / might want different abstractions.

**Pull request closed without merging**
